base/bucket_gocb.go

```diff
@@ -1053,29 +1092,152 @@ func (bucket CouchbaseBucketGoCB) GetWithXattr(k string, xattrKey string, rv int
+// This is currently called only for Purge operations.
+//
+// The doc existing doc is expected to be in one of the following states:
+//   - DocExists and XattrExists
+//   - DocExists but NoXattr
+//   - XattrExists but NoDoc
+//   - NoDoc and NoXattr
+// In all cases, the end state will be NoDoc and NoXattr.
+// Exepcted errors:
+//    - Temporary server overloaded errors, in which case the caller should retry
+//    - If the doc is in the the NoDoc and NoXattr state, it will return a KeyNotFound error
 func (bucket CouchbaseBucketGoCB) DeleteWithXattr(k string, xattrKey string) error {
 
+	// Delegate to internal method that can take a testing-related callback
+	return bucket.deleteWithXattrInternal(k, xattrKey, nil)
+
+}
+
+// A function that will be called back after the first delete attempt but before second delete attempt
+type deleteWithXattrRaceInjection func(bucket CouchbaseBucketGoCB, k string, xattrKey string)
+
+func (bucket CouchbaseBucketGoCB) deleteWithXattrInternal(k string, xattrKey string, callback deleteWithXattrRaceInjection) error {
+
 	bucket.singleOps <- struct{}{}
 	defer func() {
 		<-bucket.singleOps
 	}()
 	gocbExpvars.Add("Delete", 1)
-	removeCas, err := bucket.Bucket.Remove(k, 0)
-	if err != nil && err != gocb.ErrKeyNotFound {
+
+	LogTo("CRUD+", "DeleteWithXattr called with key: %v xattrKey: %v", k, xattrKey)
+
+	// Try to delete body and xattrs in single op
+	// NOTE: ongoing discussion w/ KV Engine team on whether this should handle cases where the body
+	// doesn't exist (eg, a tombstoned xattr doc) by just ignoring the "delete body" mutation, rather
+	// than current behavior of returning gocb.ErrKeyNotFound
+	_, mutateErr := bucket.Bucket.MutateInEx(k, gocb.SubdocDocFlagNone, gocb.Cas(0), uint32(0)).
+		RemoveEx(xattrKey, gocb.SubdocFlagXattr). // Remove the xattr
+		RemoveEx("", gocb.SubdocFlagNone).        // Delete the document body
+		Execute()
+
+	// If no error, or it was just a ErrSubDocSuccessDeleted error, we're done.
+	// ErrSubDocSuccessDeleted is a "success error" that means "operation was on a tombstoned document"
+	if mutateErr == nil || mutateErr == gocbcore.ErrSubDocSuccessDeleted {
+		LogTo("CRUD+", "No error or ErrSubDocSuccessDeleted.  We're done.")
+		return nil
+	}
+
+	switch {
+	case bucket.IsKeyNotFoundError(mutateErr):
+		// KeyNotFound indicates there is no doc body.  Try to delete only the xattr.
+		return bucket.deleteDocXattrOnly(k, xattrKey, callback)
+	case bucket.IsSubDocPathNotFound(mutateErr):
+		// KeyNotFound indicates there is no XATTR.  Try to delete only the body.
+		return bucket.deleteDocBodyOnly(k, xattrKey, callback)
+	default:
+		// return error
+		return mutateErr
+	}
+
+}
+
+func (bucket CouchbaseBucketGoCB) deleteDocXattrOnly(k string, xattrKey string, callback deleteWithXattrRaceInjection) error {
+
+	// Invoke the callback which has the ability to change the document state
+	if callback != nil {
+		callback(
+			bucket,
+			k,
+			xattrKey,
+		)
+	}
```

Comment:
I feel like we should contain the callback usage to deleteWithXattrInternal.  It doesn't really add value inside this method, since it's getting invoked first, and adds overhead for any future consumers of this method.  I feel like it's cleaner to invoke the callback within the error switch statement (in both places).